### PR TITLE
Msvc

### PR DESCRIPTION
**Pull request by aveenismail**
@klali

### CODE REVIEW

File: common/time_win.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+//
```
Comment (klali) [inline common/time_win.h:1]: Copyright and license here.

File: common/time_win.h [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+// Created by aveen on 2020-08-24.
+//
+
+#ifndef YUBIHSM_SHELL_TIME_WIN_H
+#define YUBIHSM_SHELL_TIME_WIN_H
+
+#include <time.h>
+
+int gettimeofday_win(struct timeval *__restrict tv);
```
Comment (klali) [inline common/time_win.h:10]: is the __restrict keyword needed here?

File: lib/yubihsm.h [hunk 3]
```diff
@@ -554,7 +554,8 @@ typedef struct {
   uint8_t sequence;
   /// Object origin
   uint8_t origin;
-  /// Object label. The label consists of raw bytes and is not restricted to printable characters or valid UTF-8 glyphs
+  /// Object label. The label consists of raw bytes and is not restricted to
```
Comment (klali) [inline lib/yubihsm.h:557]: While it's potentially good to let clang-format fix this it doesn't really belong with the rest of this, I wouldn't add this but push it to the future.

File: lib/yubihsm_usb.c [hunk 4]
```diff
@@ -74,7 +75,7 @@ static yh_rc backend_send_msg(yh_backend *connection, Msg *msg, Msg *response,
                               const char *identifier) {
   int32_t trf_len = msg->st.len + 3;
   yh_rc ret = YHR_GENERIC_ERROR;
-  unsigned long read_len;
+  unsigned long read_len = 0;
```
Comment (klali) [inline lib/yubihsm_usb.c:78]: Is this needed?

File: lib/yubihsm_winhttp.c [hunk 5]
```diff
@@ -309,7 +311,10 @@ static yh_rc backend_send_msg(yh_backend *connection, Msg *msg, Msg *response,
   }
 
   if (identifier != NULL && strlen(identifier) > 0 && strlen(identifier) < 32) {
-    swprintf(hsm_identifier, 128, L"YubiHSM-Session: %s", identifier);
+    tmp_identifier_len = mbstowcs(NULL, identifier, 0);
+    tmp_identifier = calloc(tmp_identifier_len + 1, sizeof(wchar_t));
+    mbstowcs(tmp_identifier, identifier, tmp_identifier_len + 1);
+    swprintf(hsm_identifier, 128, L"YubiHSM-Session: %s", tmp_identifier);
```
Comment (klali) [inline lib/yubihsm_winhttp.c:317]: Did you try this with %S or %ls instead of doing the conversion by hand? Since we know this code now only compiles correctly with MSVC should we use #error (or #warn) if this path is hit with gcc?

File: lib/yubihsm_winhttp.c [hunk 6]
```diff
@@ -102,9 +102,9 @@ static bool parseUrl(char *url, struct urlComponents *components) {
   return true;
 }
 
-static void CALLBACK http_callback(HINTERNET internet __attribute__((unused)),
-                                   DWORD_PTR context, DWORD status,
-                                   LPVOID statusInfo, DWORD statusInfoLen) {
+static void CALLBACK http_callback(HINTERNET internet, DWORD_PTR context,
```
Comment (klali) [inline lib/yubihsm_winhttp.c:105]: doesn't unused work on msvc? this will generate a new warning then I guess? We have an UNUSED macro in yubihsm.h that can be used in that case.

File: pkcs11/CMakeLists.txt [hunk 7]
```diff
@@ -75,6 +86,7 @@ target_link_libraries(
   yubihsm_pkcs11
   yubihsm
   ${LIBCRYPTO_LDFLAGS}
+  ${GETOPT_LIBS}
```
Comment (klali) [inline pkcs11/CMakeLists.txt:89]: what's this? do we need to link with a special getopt lib?

File: pkcs11/util_pkcs11.c [hunk 8]
```diff
@@ -2147,14 +2150,9 @@ CK_RV apply_sign_mechanism_finalize(yubihsm_pkcs11_op_info *op_info) {
   return CKR_OK;
 }
 
-CK_RV apply_verify_mechanism_finalize(yubihsm_pkcs11_op_info *op_info
-                                      __attribute((unused))) {
+CK_RV apply_verify_mechanism_finalize() { return CKR_OK; }
```
Comment (klali) [inline pkcs11/util_pkcs11.c:2153]: Where did the argument go here?

File: pkcs11/util_pkcs11.h [hunk 9]
```diff
@@ -62,7 +62,7 @@ bool check_verify_mechanism(yubihsm_pkcs11_slot *slot,
 CK_RV apply_verify_mechanism_init(yubihsm_pkcs11_op_info *op_info);
 CK_RV apply_verify_mechanism_update(yubihsm_pkcs11_op_info *op_info,
                                     CK_BYTE_PTR in, CK_ULONG in_len);
-CK_RV apply_verify_mechanism_finalize(yubihsm_pkcs11_op_info *op_info);
+CK_RV apply_verify_mechanism_finalize();
```
Comment (klali) [inline pkcs11/util_pkcs11.h:65]: We shouldn't drop this argument (I'm fine with discussing if the function is at all needed, but since we call it with an op_info we shouldn't declare the function like this).

File: pkcs11/util_pkcs11.c [hunk 10]
```diff
@@ -3151,7 +3149,7 @@ CK_RV parse_ec_template(CK_ATTRIBUTE_PTR pTemplate, CK_ULONG ulCount,
                         yubihsm_pkcs11_object_template *template) {
 
   uint8_t *ecparams = NULL;
-  uint16_t ecparams_len;
+  uint16_t ecparams_len = 0;
```
Comment (klali) [inline pkcs11/util_pkcs11.c:3152]: is this change needed?

File: src/main.c [hunk 11]
```diff
@@ -56,12 +53,23 @@
...
+#include <sys/time.h>
 #include <editline/readline.h>
 #include <histedit.h>
 
 History *g_hist;
 #endif
 
+#ifdef _MSVC
+//#define S_ISLNK(m) (((m) & S_IFMT) == 0120000) // todo: verify correctness:
+// https://github.com/fluent/fluent-bit/blob/f07fdab96b05bba77af2da302c379c9e3f6433f4/lib/libbacktrace-ca0de05/elf.c
```
Comment (klali) [inline src/main.c:67]: the link reference isn't very easy to follow here, maybe change it to point to the interesting line?
the commented #define with a todo, do we need it at all? is it something we want?

File: src/main.c [hunk 12]
```diff
@@ -709,9 +717,8 @@ static bool probe_session(yubihsm_context *ctx, int index) {
 }
 
 #ifdef __WIN32
-static void WINAPI timer_handler(void *lpParam __attribute__((unused)),
-                                 unsigned char TimerOrWaitFired
-                                 __attribute__((unused))) {
+static void WINAPI timer_handler(void *lpParam,
+                                 unsigned char TimerOrWaitFired) {
```
Comment (klali) [inline src/main.c:721]: We should still mark these as unused, otherwise we'll get a warning when compiling?

File: src/main.c [hunk 13]
```diff
@@ -2652,7 +2659,7 @@ int main(int argc, char *argv[]) {
     }
 
   } else {
-    int num;
+    int num = 0;
```
Comment (klali) [inline src/main.c:2662]: is this needed?

**Review by klali — CHANGES_REQUESTED** [review]

File: src/main.c (hunk 13 above)
Comment (klali) [reply src/main.c:2662]: I think it is, right?

File: lib/platform-config.h.in [hunk 14]
```diff
@@ -0,0 +1,48 @@
+/*
```
[the remark this answers is not in the document]
Comment (aveenismail) [reply lib/platform-config.h.in]: I thought about putting it in common too, actually, but ended up in lib. And you're right, common is not dependent on what is outside of it.

I moved it now to common, but that caused the MAC build on Travis to fail, so I'll need to look into that

File: common/debug.h [hunk 15]
```diff
@@ -41,6 +45,23 @@
...
   (strrchr(__FILE__, '/') ? strrchr(__FILE__, '/') + 1 : __FILE__)
 
+#ifdef _MSVC
+#define D(var, file, col, who, lev, ...)                                       \
+  if (var) {                                                                   \
+    struct timeval _tv;                                                        \
+    struct tm _tm;                                                             \
+    char _tbuf[20];                                                            \
+    time_t _tsecs;                                                             \
+    gettimeofday_win(&_tv);                                                    \
```
[the remark this answers is not in the document]
Comment (aveenismail) [reply common/debug.h]: Yes, it's only gettimeofday() and localtime_x() that differ. 

Fixed by re-defining those functions for MSVC

File: common/time_win.c [hunk 16]
```diff
@@ -0,0 +1,15 @@
+//
```
[the remark this answers is not in the document]
Comment (aveenismail) [reply common/time_win.c:1]: Fixed

File: common/time_win.h (hunk 1 above)
Comment (aveenismail) [reply common/time_win.h:1]: Fixed

File: lib/yubihsm_usb.c (hunk 4 above)
Comment (aveenismail) [reply lib/yubihsm_usb.c:78]: MSVC gave a warning because read_len was not initialized. Do you think it's better to put it back?

File: lib/yubihsm_winhttp.c (hunk 5 above)
Comment (aveenismail) [reply lib/yubihsm_winhttp.c]: Ahh... sorry, I forgot about this. Fixed how

File: lib/yubihsm_usb.c (hunk 4 above)
Comment (klali) [reply lib/yubihsm_usb.c:78]: if it warns without it leave it there. I don't think it's actually true but doesn't matter.

File: lib/yubihsm_winhttp.c (hunk 6 above)
Comment (aveenismail) [reply lib/yubihsm_winhttp.c:106]: No, that caused an error on MSVC. Replaced it now with UNUSED macro

File: pkcs11/CMakeLists.txt (hunk 7 above)
Comment (aveenismail) [reply pkcs11/CMakeLists.txt:89]: Yes. 

gengetopt is not available on Windows. Before, we used to use gengetopt binaries compiled with mingw, but this does not work with MSVC. So instead, cmdline.c/h are generated on a nix system and builds on Windows use the source release package and not the source directly cloned from the repository. getopt.dll is the "getopt"-part that's available with MSVC. It will be included in the SDK package along with its license.

File: pkcs11/util_pkcs11.c (hunk 8 above)
Comment (aveenismail) [reply pkcs11/util_pkcs11.c]: I put it back and used the UNUSED macro.

(MSVC did not like __attribute((unused)) and gnu didn't like an unused argument that wasn't marked as unused)

File: pkcs11/util_pkcs11.h (hunk 9 above)
Comment (aveenismail) [reply pkcs11/util_pkcs11.h]: Fixed

File: pkcs11/util_pkcs11.c (hunk 10 above)
Comment (aveenismail) [reply pkcs11/util_pkcs11.c:3156]: MSVC warning that the variable is not initialized

File: src/main.c (hunk 11 above)
Comment (aveenismail) [reply src/main.c]: Removed the comment and created a jira ticket for it instead. HSM-294

File: src/main.c (hunk 12 above)
Comment (aveenismail) [reply src/main.c:721]: Fixed with UNUSED macro

File: src/main.c (hunk 13 above)
Comment (aveenismail) [reply src/main.c:2663]: MSVC warning that the variable isn't initialized

File: common/time_win.h (hunk 2 above)
Comment (aveenismail) [reply common/time_win.h]: Probably not. I had a different approach at the beginning and so used the same declaration as gettimeofday(). I just didn't change the declaration enough after that. 

Removed it now

File: lib/yubihsm.h (hunk 3 above)
Comment (aveenismail) [reply lib/yubihsm.h:557]: Not sure I can do much about it without changing pre-commit setup. Commit fails without it

File: pkcs11/util_pkcs11.c [hunk 17]
```diff
@@ -1637,8 +1640,11 @@ yubihsm_pkcs11_object_desc *get_object_desc(yh_session *session,
   }
 
   object->object.type = type;
-
+#ifdef _MSVC
+  gettimeofday_win(&object->tv);
```
Comment (klali) [inline pkcs11/util_pkcs11.c:1644]: why not use the earlier defined macro to get the right implementation no matter platform?

**Review by klali — APPROVED** [review]
lgtm
